### PR TITLE
upload: add option to mask uploaded file name

### LINKED ISSUE #1247: randomize/erase filename
option to either
- set filename to meguca hash
- randomize based on some pattern

when uploading a file.

### PR DESCRIPTION
**Pull request by Kamuso**
Closes #1247 
I got it to work, but I'm not very good with postgres so the database bits might not be beautiful.

### CODE REVIEW

**Comment by bakape** [comment]
Why need any server upload code changes at all? This can be done completely client-side in `upload.ts`. Just upload with a different name.

**Comment by Kamuso** [comment]
To avoid computing the hash again. Should I change it?

**Comment by bakape** [comment]
The file name is not a part of the hash.

**Comment by Kamuso** [comment]
Yes I know. I mean to avoid computing the hash client side for the filename, and computing it server side for identification. I know the hash is computed client side to check if it's been uploaded already, but that's only on https.
Again, do you want me to change it?

File: db/images.go [hunk 1]
```diff
@@ -148,18 +148,20 @@ func HasImage(id uint64) (has bool, err error) {
 	return
 }
 
-// InsertImage insert and image into and existing open post and return image
+// InsertImage insert an image into an existing open post and return image
 // JSON
-func InsertImage(tx *sql.Tx, postID uint64, token, name string, spoiler bool,
+func InsertImage(
+	tx *sql.Tx, postID uint64, token, name string, spoiler, mask bool,
```
Comment (bakape) [inline db/images.go:154]: The way I se see it, there is no need to alter this function. Why would we need to store, if the image was masked? Just have the client send the already masked name.

File: db/migrations.go [hunk 2]
```diff
@@ -1521,6 +1521,12 @@ var migrations = []func(*sql.Tx) error{
 			`alter table posts alter column flag type varchar(5)`,
 		)
 	},
+	func(tx *sql.Tx) (err error) {
```
Comment (bakape) [inline db/migrations.go:1524]: Meaning this is not needed.

File: static/src/sql/functions/insert_image.sql [hunk 3]
```diff
@@ -1,27 +1,35 @@
 -- Inserts image into existing post and return image json
```
Comment (bakape) [inline static/src/sql/functions/insert_image.sql:1]: Nor any changes here.

File: websockets/post_creation.go [hunk 4]
```diff
@@ -40,8 +40,8 @@ type ReplyCreationRequest struct {
 
 // ImageRequest contains data for allocating an image
 type ImageRequest struct {
-	Spoiler     bool
-	Token, Name string
+	Spoiler, Mask bool
```
Comment (bakape) [inline websockets/post_creation.go:43]: Nor here.

**Review by bakape — CHANGES_REQUESTED** [review]
Basically, yes. Please keep it completely client side (templates rendered on the server are the client side).

File: db/images.go (hunk 1 above)
Comment (Kamuso) [reply db/images.go]: >Why would we need to store, if the image was masked?

It wasn't stored. It was used for some logic in the sql query.

**Comment by Kamuso** [comment]
Okay, finished rewriting.

**Comment by Kamuso** [comment]
Wait a minute, some of the files didn't revert properly.

**Comment by Kamuso** [comment]
Okay, now it's fine.

File: client/posts/posting/upload.ts [hunk 5]
```diff
@@ -340,3 +363,57 @@ function bufferToHex(buf: ArrayBuffer): string {
...
+}
+
+export function initUpload() {
+    // Add event to optionally mask uploaded file's name when creating
+    // a new thread
+    if (!page.thread) {
+        const form = document.
+            getElementById("new-thread-form") as HTMLFormElement;
+        const mask = form.
+            querySelector("input[name=mask") as HTMLInputElement;
```
Comment (bakape) [inline client/posts/posting/upload.ts:406]:
```suggestion
            querySelector("input[name=mask]") as HTMLInputElement;
```

**Review by bakape — CHANGES_REQUESTED** [review]

**Comment by bakape** [comment]
Okay, thanks!

**Pull request merged**